Generation/src/approaches/random/Generator.java

```diff
@@ -279,18 +284,67 @@ else if (rule.lhs().equals("<dim>"))
 			return "";
 		}
 
-		final EBNFClause clause = rule.rhs().get(rng.nextInt(rule.rhs().size()));
+		// We must sort the clauses to ensure that multiple iterations, with the same rng seed, produce the same output
+		List<EBNFClause> clauses = new LinkedList<>(rule.rhs());
+		Collections.sort(clauses, Comparator.comparing(EBNFClause::toString));
+		final EBNFClause clause = clauses.get(rng.nextInt(rule.rhs().size()));
 		
 //		System.out.println("clause: " + clause + " " + (clause.isTerminal() ? "T" : "") +  (clause.isRule() ? "R" : "") +  (clause.isConstructor() ? "C" : ""));
 		
 		if (clause.isTerminal())
 		{
+
```

Comment:
Right before this I'm sorting based on clause.toString(). Meaning that I'm calling the method many more times than once per if-block, and I'm only doing it for consistency across different runs.

Assuming clauses are immutable, would it be okay if we cached the result of the toString method in the EBNFClause class? Also, do you think there's any way we could guarantee the ordering of rule.rhs() across different runs without performance penalties?

Reply:
That's true. But the few repeated calls you do yourself are very easy to fix. Fixing it also for the calls that originate from the sorting is a bit more annoying. Personally, I'd be fine with only optimising the easy case for now, and only looking at the other stuff later if it turns out that this code is actually performance-critical, and someone ends up getting annoyed by it being too slow.

Indeed, as you suggest, caching results of `toString()` calls inside `EBNFClause` objects could be one way to optimise it also for the sorting. I imagine it'd *probably* be fine, but I'm not 100% sure. The only potential concern is that these objects are not just being called during random-game-generation, but also in essentially *all* other Ludii code we have, because we basically always compile games. I'm not 100% sure if increased memory usage there might turn out to be problematic (I suspect not, just not sure).

Another solution could be for you to create a little wrapper class here, that wraps around the `EBNFClause` objects + their strings, and sort a list of such objects. Then you do the string caching only here for this purpose, and not across the rest of Ludii. But that's a bit more effort, need to write an extra little class then.

I don't see any other obvious ways of guaranteeing reproducibility.

Reply:
Sounds good. I removed the toString calls in the if-blocks. For now I'm also in favour of avoiding micro optimisations, we'll see if we need them later. However, I'd like to make my code a little more readable. Do you think we could raise the language level? The TrevisCI bot doesn't seem to like my string switch statements or the string.repeat method.

Reply:
We've considered raising the minimum supported version of Java before, but no it's not something we want to do. If I recall correctly, it was related to licensing issues for Java versions beyond 8 (from Oracle). Something about Java 8 still being easy to download for users, but later versions requiring them to first create an Oracle account and accept a license, or something like that. Of course there's always also OpenJDK, but Ludii also has a significant non-programmers userbase, and they download Java by clicking the first thing that shows up in a google search (which is Oracle).

Reply:
That's unfortunate, but it makes sense. I think th PR is good to go now.